### ISSUE TITLE
Publish prereleases (e.g. `v0.15.0-alpha-05`) to NPM, not just GitHub

### ISSUE THREAD

**Issue by JordanMartinez** [purescript/purescript#4294]
## Summary

Whenever we make prereleases that include breaking changes, publish them on NPM via the `next` tag (e.g. `npm i purescript@next`).

## Motivation

Before we can make a breaking change release (e.g. `0.15.0`), we have to update at least the core libraries to compile on and and build tools to compile with the upcoming compiler version. So, we made prereleases by publishing them on GitHub but not on NPM (e.g. `v0.14.0-rc1`, `v0.15.0-alpha-05`, etc.). Each project's CI had to be updated to download this GitHub release. 

For libraries, this is no longer a problem. During this `0.15.0` release/migration work, we've switched to using [`setup-purescript`](https://github.com/purescript-contrib/setup-purescript) to get the prerelease. However, this tool only works because it was updated to do that during the `0.15.0` work.

For build tools, this is still a bit of a problem. While we could use `setup-purescript` to download the prerelease, it's more awkward to use. `pulp` and `spago` may be testing their code against multiple versions of the `purs` binary and against multiple OSes. So, using a custom download script or multiple `setup-purescript` steps rather than a build matrix where each build is testing one version is awkward. It would be easier if `npm i -g purescript@next` could be used instead.

Finally, publishing the prerelease on NPM would allow others to use it early and see if there are any bugs that haven't been reported yet. During the `0.15.0` work, a few bugs were found by people testing it out:
- #4274
- https://github.com/purescript-web/purescript-web-html/pull/75

## Proposal

Fixing this would require updating the `RELEASE_GUIDE.md` to further explain that one should publish a pre-release on NPM (besides the other updates we should do for that file). Specifically, the below line in the file...
> After all of the prebuilt binaries are present on the GitHub releases page, publish to npm: change to the npm-package directory and run npm publish. It's a good idea to check that the package can be installed from npm at this point.
should be updated to say something like, "When making a prerelease, change to the `npm-package` directory and run `npm publish --tag next`.

See https://manpages.org/npm-publish

**Comment by JordanMartinez** [comment]
> Can I suggest to automatically release as next on every merge (independent of version updates etc)? That could make testing any new feature easier and could be automated.

I think that makes sense.

**Comment by JordanMartinez** [comment]
As far as core team members go, the following seem to be in agreement with this idea:
- myself
- Gary (via the :+1: above)
- Fabrizio (who initially suggested this)
- Ryan (via the :+1: above)

No other core team members have brought reasons against this. I'll still leave this open for a few days in case others are busy and missed this issue, but for now I will create a PR that updates the language.

**Comment by thomashoneyman** [comment]
I'm in agreement with this idea as well. I'm not sure about automatically releasing as next on every merge (probably this is fine, but I haven't thought through the implications), but I'm certainly on board with releasing pre-releases as `@next`.

**Comment by rhendric** [comment]
For the record, I'm actually rather more enthusiastic about continuous deployment than I am about extending the manual release process. I wouldn't veto manually publishing the prereleases with our release wrangler in favor of it, but the advantages of:
* having a token that can publish to npm stored in the infrastructure instead of relying on humans having available and operational dev machines in order to publish
* procedurally establishing that `next` is bleeding-edge and humans don't pre-vet what gets published to that tag—it can have lots of oopsie releases in it and they won't force us to skip any semantic version numbers that we then have to explain to users
* making it possible to test, in the same environment in which it would be used by end users, the state of the compiler immediately before bumping the version number in a semantic release, so hopefully oopsie releases in the `latest` tag and on GitHub go extinct altogether
* making arbitrary PRs immediately and easily testable/usable by interested and adventurous users, once merged, instead of making everyone wait for our release cycle to roll around or build from source

are what made this worth an explicit :+1: from me.

**Comment by JordanMartinez** [comment]
> I'm actually rather more enthusiastic about continuous deployment than I am about extending the manual release process

I'm on board for the continuous development!

**Comment by JordanMartinez** [comment]
Since we have 5 core team members in agreement with this, I think this is ready to move forward.

**Comment by JordanMartinez** [comment]
Just published the prerelease on NPM and then tried to install it locally via `npm i purescript@next`. I got this error:
```
$ npm i purescript@next
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm ERR! code 1
npm ERR! path /home/jordan/Programming/Projects/za/node_modules/purescript
npm ERR! command failed
npm ERR! command sh -c install-purescript --purs-ver=0.15.0-alpha-06
npm ERR! [ FAILURE ] Check if a prebuilt 0.15.0-alpha-06 binary is provided for linux
npm ERR! Error: Expected `version` option to be a string of PureScript version, for example '0.12.5', but got an invalid version '0.15.0-alpha-06'.
npm ERR!     at /home/jordan/Programming/Projects/za/node_modules/purescript-installer/download-purescript/index.js:97:20
npm ERR!     at new Subscription (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:197:34)
npm ERR!     at Observable.subscribe (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:279:14)
npm ERR!     at /home/jordan/Programming/Projects/za/node_modules/purescript-installer/download-or-build-purescript/index.js:266:6
npm ERR!     at new Subscription (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:197:34)
npm ERR!     at Observable.subscribe (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:279:14)
npm ERR!     at /home/jordan/Programming/Projects/za/node_modules/purescript-installer/install-purescript/index.js:133:62
npm ERR!     at new Subscription (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:197:34)
npm ERR!     at Observable.subscribe (/home/jordan/Programming/Projects/za/node_modules/zen-observable/lib/Observable.js:279:14)
npm ERR!     at Object.<anonymous> (/home/jordan/Programming/Projects/za/node_modules/purescript-installer/index.js:370:4)
npm ERR!     at Module._compile (node:internal/modules/cjs/loader:1101:14)
npm ERR!     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
npm ERR!     at Module.load (node:internal/modules/cjs/loader:981:32)
npm ERR!     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
npm ERR!     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
npm ERR!     at node:internal/main/run_main_module:17:47
```

**Comment by JordanMartinez** [comment]
I'm guessing we'll need to tweak our `npm-installer` tool: https://github.com/purescript/npm-installer

**Comment by JordanMartinez** [comment]
The error above is caused by a validation test failing. `npm-installer` uses simple regex test that doesn't account for prereleases or build metadata. So, I've opened https://github.com/purescript/npm-installer/pull/34 to use semver to that version test. 

Since I'm not sure who is watching that repo, can I get an approval?

**Comment by JordanMartinez** [comment]
Looks like I'll need to get publishing rights for `install-purescript` first.

**Comment by JordanMartinez** [comment]
The alpha-07 release can be installed via `npm i purescript@next` now.

**Comment by f-f** [comment]
Should we reopen this issue to track the continuous deployment to NPM?

**Comment by JordanMartinez** [comment]
I don't think so. Rather, I think that should be its own separate issue. I'll open a new one.

**Comment by JordanMartinez** [comment]
@rhendric Would you mind opening an issue proposing this?

**Comment by rhendric** [comment]
Will do, but it might be a few days before I can write up a full proposal.
